File: projects/basis/sdk-js/src/modules/PredictionMarkets.ts
import { BasisClient } from '../BasisClient';
import AMarketTradingArtifact from '../abis/AMarketTrading.json';
import ATokenFactoryArtifact from '../abis/ATokenFactory.json';
import IERC20Artifact from '../abis/IERC20.json';
import { Address, getAddress, keccak256, toBytes } from 'viem';

export class PredictionMarketsModule {
  private client: BasisClient;
  private marketTradingAddress: Address;

  constructor(client: BasisClient, marketTradingAddress: Address) {
    this.client = client;
    this.marketTradingAddress = marketTradingAddress;
  }

  /**
   * Helper to approve tokens for the MarketTrading contract
   */
  private async approveIfNeeded(tokenAddress: Address, amount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required for approval.");
    }
    
    const account = this.client.walletClient.account;

    // Check allowance
    const allowance = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: IERC20Artifact.abi,
      functionName: 'allowance',
      args: [account.address, this.marketTradingAddress],
    }) as bigint;

    if (allowance < amount) {
      const { request } = await this.client.publicClient.simulateContract({
        account,
        address: tokenAddress,
        abi: IERC20Artifact.abi,
        functionName: 'approve',
        args: [this.marketTradingAddress, amount],
      });

      const hash = await this.client.walletClient.writeContract(request);
      await this.client.publicClient.waitForTransactionReceipt({ hash });
    }
  }

  /**
   * Creates a new prediction market.
   */
  async createMarket(
    marketName: string,
    symbol: string,
    endTime: bigint,
    optionNames: string[],
    maintoken: Address,
    frozen: boolean,
    bonding: bigint,
    seedAmount: bigint = 0n
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Read the ecosystem's factory address, then fetch the fee
    const ecoData = await this.client.publicClient.readContract({
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'ecosystems',
      args: [maintoken],
    }) as any;
    const factoryAddress = ecoData.factory ?? ecoData[0];
    const feeAmount = await this.client.publicClient.readContract({
      address: factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'feeAmount',
    }) as bigint;

    // Auto-approve USDB for seed amount if needed
    if (seedAmount > 0n) {
      await this.approveIfNeeded(this.client.usdbAddress, seedAmount);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'createMarket',
      args: [marketName, symbol, endTime, optionNames, maintoken, frozen, bonding, seedAmount],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    return { hash, receipt };
  }

  /**
   * Creates a prediction market and registers its metadata on IPFS in one call.
   * Requires SIWE authentication.
   *
   * Returns { hash, receipt, marketTokenAddress, imageUrl, metadata }
   */
  async createMarketWithMetadata(options: {
    marketName: string;
    symbol: string;
    endTime: bigint;
    optionNames: string[];
    maintoken: Address;
    frozen?: boolean;
    bonding?: bigint;
    seedAmount?: bigint;
    description?: string;
    imageUrl?: string;
    website?: string;
    telegram?: string;
    twitterx?: string;
  }) {
    const createResult = await this.createMarket(
      options.marketName,
      options.symbol,
      options.endTime,
      options.optionNames,
      options.maintoken,
      options.frozen ?? false,
      options.bonding ?? 0n,
      options.seedAmount ?? 0n,
    );

    if (createResult.receipt.status === 'reverted') {
      throw new Error(`Market creation reverted (tx: ${createResult.hash})`);
    }

    // Parse market token from MarketCreated event
    const MARKET_CREATED_TOPIC = keccak256(toBytes('MarketCreated(address,address,address)'));
    const marketLog = createResult.receipt.logs.find(
      (l: any) => l.address.toLowerCase() === this.marketTradingAddress.toLowerCase() && l.topics[0] === MARKET_CREATED_TOPIC
    );

    let marketTokenAddress: Address;
    if (marketLog && marketLog.topics[1]) {
      marketTokenAddress = getAddress('0x' + marketLog.topics[1].slice(26)) as Address;
    } else {
      throw new Error('Could not extract market address from creation logs.');
    }

    // Upload image if provided
    let imageUrl: string | undefined;
    if (options.imageUrl) {
      imageUrl = await this.client.api.uploadImageFromUrl(options.imageUrl, marketTokenAddress);
    }

    // Create metadata
    const metadata = await this.client.api.updateMetadata({
      address: marketTokenAddress,
      description: options.description,
      image: imageUrl,
      website: options.website,
      telegram: options.telegram,
      twitterx: options.twitterx,
    });

    return {
      hash: createResult.hash,
      receipt: createResult.receipt,
      marketTokenAddress,
      imageUrl,
      metadata,
    };
  }

  /**
   * Executes an AMM buy for a prediction outcome.
   */
  async buy(
    marketToken: Address,
    outcomeId: number,
    inputToken: Address,
    inputAmount: bigint,
    minUsdb: bigint,
    minShares: bigint
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    await this.approveIfNeeded(inputToken, inputAmount);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'buy',
      args: [marketToken, outcomeId, inputToken, inputAmount, minUsdb, minShares],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    return { hash, receipt };
  }

  /**
   * Claims winnings after a market resolves.
   */
  async redeem(marketToken: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'redeem',
      args: [marketToken],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    return { hash, receipt };
  }

  /**
   * Reads the MarketData struct.
   */
  async getMarketData(marketToken: Address) {
    return this.client.publicClient.readContract({
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'getMarketData',
      args: [marketToken],
    });
  }

  /**
   * Reads the Outcome struct.
   */
  async getOutcome(marketToken: Address, outcomeId: number) {
    return this.client.publicClient.readContract({
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'getOutcome',
      args: [marketToken, outcomeId],
    });
  }

  /**
   * Reads user balances.
   */
  async getUserShares(marketToken: Address, user: Address, outcomeId: number) {
    return this.client.publicClient.readContract({
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'getUserShares',
      args: [marketToken, user, outcomeId],
    });
  }

  /**
   * Returns the initial reserves required for a given number of outcomes.
   */
  async getInitialReserves(numOutcomes: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'getInitialReserves',
      args: [numOutcomes],
    }) as Promise<bigint>;
  }

  /**
   * Buys from order book and AMM in a single transaction.
   * Fills specified orders first, then routes remaining input to the AMM.
   */
  async buyOrdersAndContract(
    marketToken: Address,
    outcomeId: number,
    orderIds: bigint[],
    inputToken: Address,
    totalInput: bigint,
    minShares: bigint
  ) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    await this.approveIfNeeded(inputToken, totalInput);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.marketTradingAddress,
      abi: AMarketTradingArtifact.abi,
      functionName: 'buyOrdersAndContract',
      args: [marketToken, outcomeId, orderIds, inputToken, totalInput, minShares],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    return { hash, receipt };
  }
}

File: projects/basis/basis-sdk-python/basis/modules/prediction_markets.py
from web3 import Web3
from .factory import load_abi

class PredictionMarketsModule:
    def __init__(self, client, market_trading_address: str):
        self.client = client
        self.market_trading_address = Web3.to_checksum_address(market_trading_address)
        self.market_trading_abi = load_abi('AMarketTrading.json')
        self.erc20_abi = load_abi('IERC20.json')
        self.contract = self.client.web3.eth.contract(address=self.market_trading_address, abi=self.market_trading_abi)

    def _approve_if_needed(self, token_address: str, amount: int):
        if not self.client.account:
            raise ValueError("Wallet account is required for approval.")

        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.erc20_abi)
        
        allowance = token_contract.functions.allowance(
            self.client.account.address, self.market_trading_address
        ).call()

        if allowance < amount:
            func = token_contract.functions.approve(self.market_trading_address, amount)
            tx = func.build_transaction({
                'from': self.client.account.address,
                'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            })
            signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
            tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
            self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

    def _build_and_send_tx(self, function_call):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        
        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
        })

        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        
        return {
            'hash': tx_hash.hex(),
            'receipt': receipt
        }

    def _create_market(self, market_name: str, symbol: str, end_time: int, option_names: list[str], maintoken: str, frozen: bool, bonding: int, seed_amount: int = 0):
        """Internal: creates a market on-chain. Use create_market_with_metadata(). Auto-fetches and attaches the creation fee."""
        checksum_maintoken = Web3.to_checksum_address(maintoken)
        # Read ecosystem's factory address, then fetch fee
        eco_data = self.contract.functions.ecosystems(checksum_maintoken).call()
        factory_address = eco_data[0]  # factory is first field in Ecosystem struct
        factory_abi = load_abi('ATokenFactory.json')
        factory_contract = self.client.web3.eth.contract(address=factory_address, abi=factory_abi)
        fee_amount = factory_contract.functions.feeAmount().call()

        # Auto-approve USDB for seed amount if needed
        if seed_amount > 0:
            self._approve_if_needed(self.client.usdb_address, seed_amount)

        func = self.contract.functions.createMarket(market_name, symbol, end_time, option_names, checksum_maintoken, frozen, bonding, seed_amount)
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        tx = func.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            'value': fee_amount,
        })
        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        return {'hash': tx_hash.hex(), 'receipt': receipt}

    def create_market_with_metadata(
        self,
        market_name: str,
        symbol: str,
        end_time: int,
        option_names: list[str],
        maintoken: str,
        description: str = None,
        image_url: str = None,
        website: str = None,
        telegram: str = None,
        twitterx: str = None,
        frozen: bool = False,
        bonding: int = 0,
        seed_amount: int = 0,
    ):
        """Creates a prediction market and registers metadata on IPFS in one call.

        Requires SIWE authentication.
        Returns dict with hash, receipt, market_token_address, image_url, metadata.
        """
        create_result = self._create_market(
            market_name, symbol, end_time, option_names,
            maintoken, frozen, bonding, seed_amount,
        )

        receipt = create_result['receipt']
        if receipt.get('status') == 0:
            raise RuntimeError(f"Market creation reverted (tx: {create_result['hash']})")

        # Parse market token from MarketCreated event
        market_created_topic = Web3.keccak(text="MarketCreated(address,address,address)").hex()
        mt_lower = self.market_trading_address.lower()
        market_token_address = None
        for log_entry in receipt.get('logs', []):
            addr = log_entry.get('address', '')
            if addr.lower() != mt_lower:
                continue
            topics = log_entry.get('topics', [])
            if not topics:
                continue
            t0 = topics[0].hex() if isinstance(topics[0], bytes) else str(topics[0])
            if t0 == market_created_topic and len(topics) > 1:
                raw = topics[1].hex() if isinstance(topics[1], bytes) else str(topics[1])
                market_token_address = Web3.to_checksum_address("0x" + raw[-40:])
                break

        if not market_token_address:
            raise RuntimeError("Could not extract market address from creation logs.")

        # Upload image if provided
        uploaded_image_url = None
        if image_url:
            uploaded_image_url = self.client.api.upload_image_from_url(image_url, contract_address=market_token_address)

        # Create metadata
        metadata = self.client.api.update_metadata(
            address=market_token_address,
            description=description,
            image=uploaded_image_url,
            website=website,
            telegram=telegram,
            twitterx=twitterx,
        )

        return {
            'hash': create_result['hash'],
            'receipt': receipt,
            'market_token_address': market_token_address,
            'image_url': uploaded_image_url,
            'metadata': metadata,
        }

    def buy(self, market_token: str, outcome_id: int, input_token: str, input_amount: int, min_usdb: int, min_shares: int):
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_input = Web3.to_checksum_address(input_token)
        
        self._approve_if_needed(checksum_input, input_amount)
        
        func = self.contract.functions.buy(checksum_market, outcome_id, checksum_input, input_amount, min_usdb, min_shares)
        return self._build_and_send_tx(func)

    def redeem(self, market_token: str):
        checksum_market = Web3.to_checksum_address(market_token)
        func = self.contract.functions.redeem(checksum_market)
        return self._build_and_send_tx(func)

    def get_market_data(self, market_token: str):
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getMarketData(checksum_market).call()

    def get_outcome(self, market_token: str, outcome_id: int):
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getOutcome(checksum_market, outcome_id).call()

    def get_user_shares(self, market_token: str, user: str, outcome_id: int):
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_user = Web3.to_checksum_address(user)
        return self.contract.functions.getUserShares(checksum_market, checksum_user, outcome_id).call()

    def buy_orders_and_contract(self, market_token: str, outcome_id: int, order_ids: list[int], input_token: str, total_input: int, min_shares: int):
        """Buys from order book and AMM in a single transaction. Auto-approves input token."""
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_input = Web3.to_checksum_address(input_token)
        self._approve_if_needed(checksum_input, total_input)
        func = self.contract.functions.buyOrdersAndContract(checksum_market, outcome_id, order_ids, checksum_input, total_input, min_shares)
        return self._build_and_send_tx(func)

    def get_initial_reserves(self, num_outcomes: int) -> tuple:
        """Returns (perOutcome, totalReserve) for a given number of outcomes."""
        return self.contract.functions.getInitialReserves(num_outcomes).call()

    def get_num_outcomes(self, market_token: str) -> int:
        return self.contract.functions.getNumOutcomes(Web3.to_checksum_address(market_token)).call()

    def get_option_names(self, market_token: str) -> list:
        return self.contract.functions.getOptionNames(Web3.to_checksum_address(market_token)).call()

    def has_betted_on_market(self, market_token: str, user: str) -> bool:
        return self.contract.functions.hasBettedOnMarket(
            Web3.to_checksum_address(market_token), Web3.to_checksum_address(user)
        ).call()

    def get_bounty_pool(self, market_token: str) -> int:
        return self.contract.functions.getBountyPool(Web3.to_checksum_address(market_token)).call()

    def get_general_pot(self, market_token: str) -> int:
        return self.contract.functions.getGeneralPot(Web3.to_checksum_address(market_token)).call()

    def get_buy_order_amounts_out(self, market_token: str, order_id: int, usdb_amount: int):
        return self.contract.functions.getBuyOrderAmountsOut(
            Web3.to_checksum_address(market_token), order_id, usdb_amount
        ).call()
